swarm/src/connection.rs

```diff
@@ -344,19 +349,19 @@ where
                 }
             }
 
-            // Compute new shutdown
-            if let Some(new_shutdown) =
-                compute_new_shutdown(handler.connection_keep_alive(), shutdown, *idle_timeout)
-            {
-                *shutdown = new_shutdown;
-            }
-
             // Check if the connection (and handler) should be shut down.
-            // As long as we're still negotiating substreams, shutdown is always postponed.
+            // As long as we're still negotiating substreams or have any active streams shutdown is always postponed.
             if negotiating_in.is_empty()
                 && negotiating_out.is_empty()
                 && requested_substreams.is_empty()
```

Comment:
Either of these being false, e.g. `negotiating_in` being **not** empty, implies progress, right? Should we add an `else` that sets `shutdown = Shutdown::None`?

Otherwise the idle timer (`Shutdown::Later`) is ticking even though the connection wasn't idle throughout the entire lifetime of the timer.

(This wasn't relevant before, as there wasn't a default implementation on `ConnectionHandler::connection_keep_alive` that defaulted to `false`. In other words, `shutdown` was set to `Shutdown::None` on progress.)

Reply:
This makes sense to me. We should set `shutdown = Shutdown::None` while it isn't idle.